### DISCUSSION TITLE
Attempt to re-create user with LDAP

### DISCUSSION THREAD

**Discussion by Mazakarini** [kimai/kimai#4571] [Questions & Answers]
### Describe the issue

After successful user authentication via LDAP / AD kimai, running on Ubuntu 22.04.3 LTS, tries to insert the user again in kimai2_users table, which leads to a constraint violation.

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.9.0

### How do you run Kimai?

Virtual Server or alike

### Which PHP version are you using?

8.1

### Logfile

[2024-01-16 17:45:00] [request] INFO: Matched route "security_check". {"route":"security_check","route_parameters":{"_route":"security_check","_controller":"App\\Controller\\Security\\SecurityController::checkAction","_locale":"de"},"request_uri":"https://kimai.myserver.local/de/login_check","method":"POST"} []
[2024-01-16 17:45:00] [security] INFO: Authenticator successful! {"token":{"Symfony\\Component\\Security\\Core\\Authentication\\Token\\UsernamePasswordToken":"UsernamePasswordToken(user=\"myusername\", roles=\"ROLE_TEAMLEAD, ROLE_USER\")"},"authenticator":"App\\Ldap\\LdapAuthenticator"} []
[2024-01-16 17:45:00] [cache] INFO: Lock acquired, now computing item "permissions" {"key":"permissions"} []
[2024-01-16 17:45:00] [request] CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\UniqueConstraintViolationException: "An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'myusername' for key 'kimai2_users.UNIQ_B9AC5BCEF85E0677'" at ExceptionConverter.php line 62 {"exception":"[object] (Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException(code: 1062): An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'myusername' for key 'kimai2_users.UNIQ_B9AC5BCEF85E0677' at /var/www/html/kimai2/vendor/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php:62)\n[previous exception] [object] (Doctrine\\DBAL\\Driver\\PDO\\Exception(code: 1062): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'myusername' for key 'kimai2_users.UNIQ_B9AC5BCEF85E0677' at /var/www/html/kimai2/vendor/doctrine/dbal/src/Driver/PDO/Exception.php:28)\n[previous exception] [object] (PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'myusername' for key 'kimai2_users.UNIQ_B9AC5BCEF85E0677' at /var/www/html/kimai2/vendor/doctrine/dbal/src/Driver/PDO/Statement.php:130)"} []

### Screenshots

local.yaml:
security:
    providers:
        chain_provider:
            chain:
                providers: [kimai_ldap]

kimai:
    ldap:
        activate: true
        # more infos about the connection params can be found at:
        # https://docs.laminas.dev/laminas-ldap/api/
        connection:
            # The default hostname of the LDAP server (mandatory setting).
            # You can connect to multiple servers by setting their URLs like this:
            # host: "ldap://ldap.example.local ldap://ldap2.example.local"
            # host: "ldaps://ldap.example.local ldaps://ldap2.example.local"
            host: myserver.local

            # Default port for your LDAP port server
            # default: 389
            port: 636

            # Whether or not the LDAP client should use SSL encrypted transport.
            # The useSsl and useStartTls options are mutually exclusive.
            # default: false
            useSsl: true

            # Enable TLS negotiation (should be favoured over useSsl).
            # The useSsl and useStartTls options are mutually exclusive.
            # default: false
            #useStartTls: true

            # The default credentials username (your service account). Some servers
            # require that this is given in DN form.
            # It must be given in DN form if the LDAP server requires
            # a DN to bind and binding should be possible with simple usernames.
            # default: empty
            username: CN=appuser,OU=users,OU=Users,DC=myserver,DC=local

            # Password for the username (service-account) above
            # default: empty
            password: somePwd

            # LDAP search filter to find the user (%s will be replaced by the username).
            # Should be set, to be compatible with your object structure.
            # You don not need to set this filter, unless you have a very special setup
            # or use Microsofts Active directory.
            #
            # Defaults:
            # - if bindRequiresDn is false: (&(objectClass=user)(sAMAccountName=%s))
            # - if bindRequiresDn is true: (&%filter%(uid=%s))
            #   - %filter% = empty
            #     accountFilterFormat = (&(usernameAttribute=%s))
            #   - %filter% = (&(objectClass=posixAccount))
            #     accountFilterFormat = (&(objectClass=posixAccount))(&(usernameAttribute=%s))
            #
            # %filter% is the "filter" configuration defined below in the "user" section
            accountFilterFormat: (&(objectClass=user)(sAMAccountName=%s))

            # If true, this instructs Kimai to retrieve the DN for the account,
            # used to bind if the username is not already in DN form.
            # default: true
            bindRequiresDn: true

            # If set to true, this option indicates to the LDAP client that
            # referrals should be followed, default: false
            #optReferrals: false

            # for the next options please refer to:
            # https://docs.laminas.dev/laminas-ldap/api/
            #allowEmptyPassword: false
            #tryUsernameSplit:
            #networkTimeout:
            #accountCanonicalForm: 3
            accountDomainName: myserver.local
            accountDomainNameShort: myserver

        user:
            # baseDn to query for users (mandatory setting).
            baseDn: DC=myserver,DC=local

            # Field used to match the login username in your LDAP.
            # If "bindRequiresDn: false" is set, the username is used in "bind".
            # Otherwise a search is executed to find the users "dn" by finding the user
            # via this attribute with his "baseDn" and the "filter" below.
            # default: uid
            usernameAttribute: samaccountname

            # LDAP search base filter to find the user / the users DN.
            # Do NOT include the rule (&(usernameAttribute=%s)), it will be appended
            # automatically. The result of the search filter must return 1 result only.
            # default: empty (results in (&(uid=%s)) with default usernameAttribute)
            filter: (objectClass=user)

            # LDAP search base filter to find the user attributes.
            # This is used for a slightly different query than the one above, which is
            # used to query the users DN only.
            # AD users might have too many results (Exchange activesync devices
            # attributes) and therefor an incompatible result structure if not changed.
            # See https://github.com/kevinpapst/kimai2/issues/875
            # default: (objectClass=*)
            attributesFilter: (objectClass=user)

            # Configure the mapping between LDAP attributes and user entity
            # The ldap_attr must be given in lowercase!
            attributes:
                # The following 2 rules are automatically prepended and can be overwritten.
                # Username is set to the value of the configured "usernameAttribute" field
                - { ldap_attr: sAMAccountName, user_method: setUsername }
                # Only applied if you don't configure a mapping for setEmail()
                - { ldap_attr: mail, user_method: setEmail }
                # An example which will set the display name in Kimai from the
                # value of the "common name" field in your LDAP
                - { ldap_attr: cn, user_method: setAlias }

        # You can comment the following section, if you don't want to manage
        # user roles in Kimai via LDAP groups. If you want to use the group
        # sync, you have to set at least the "role.baseDn" config.
        # default: deactivated as "role.baseDn" is empty by default
        role:
            # baseDn to query for groups, MUST be set to activate the "group import"
            # default: empty (deactivated)
            baseDn: OU=groups,OU=Security Groups,DC=myserver,DC=local

            # Filter to query user groups, all results will be matched against
            # the configured "groups" mapping below.
            # The full search filter will ALWAYS be generated like this:
            # (&%filter(userDnAttribute=valueOfUsernameAttribute))
            # The following example rule will be expanded to (for user "foo"):
            # (&(&(objectClass=groupOfNames))(member=foo))
            # default: empty
            filter: (&(objectClass=group))

            # The following field is taken from the LDAP user entry and its
            # value is used in the filter above as "valueOfUsernameAttribute".
            # The attribute must be given in lowercase!
            # The example below uses "posix group style memberUid".
            # default: dn
            #usernameAttribute: uid

            # Field that holds the group name, which will be used to map the
            # LDAP groups with Kimai roles (see groups mapping below).
            # default: cn
            #nameAttribute: cn

            # Field that holds the users dn in your LDAP group definition.
            # Value of this configuration is used in the filter (see above).
            # default: member
            #userDnAttribute: member

            # Convert LDAP group name (nameAttribute) to Kimai role
            # You will very likely have to define mappings, unless your groups
            # are called "teamlead", "admin" or "super_admin"
            groups:
                - { ldap_value: ORG_PM, role: ROLE_TEAMLEAD }
            #    - { ldap_value: kimai_admin, role: ROLE_ADMIN }

**Comment by kevinpapst** [comment]
This is NOT a bug, but a mis-configuration. Kimai cannot detect that the user already exists, because something is wrong with the DN

**Reply by kevinpapst** [reply]
And as always: make the shortest config possible. 
do not copy & paste everything you find in the docs

**Reply by Mazakarini** [reply]
FYI same settings in local.yaml work with Kimai V 1.30.7 stable (Ayumi).
What is the criteria the username? This is taken from the users sAMAccountName attribute and inserted in kimai2_users table correctly during initial login of the user.

**Reply by kevinpapst** [reply]
> FYI same settings in local.yaml work with Kimai V 1.30.7 stable (Ayumi).

I heard that argument before... and sometimes it is true. But too many user are already on 2 to prove that LDAP is working (my test setup and multiple customer environments included).

I cannot test your exact setup, so all you can do is start to check the queries:
https://www.kimai.org/documentation/ldap.html#troubleshooting--logging

And later one the code: 
https://github.com/kimai/kimai/tree/main/src/Ldap

**Reply by dodophoenix** [reply]
Hi, I've got the same problem here and did some debugging. 
Found out that there is no select query against users table during login.

It looks like the problem comes from the user entity that is new'ed in LdapManager:138
https://github.com/kimai/kimai/blob/e293d62aa1f7a87bab354c82d54c257959f43f53/src/Ldap/LdapManager.php#L138
This new entity is used during the login process. During the process the last:login time stamp is updated which causes the entityManager to create an insert query due to the missing Id of the entity.

Fetching the entity from DB using userRepository->loadUserByIdentifier() before hydrating the LDAP attributes makes it work.
